types/lib/model.d.ts

```diff
@@ -2654,8 +2654,15 @@ export abstract class Model<T = any, T2 = any> extends Hooks {
   /**
    * This is the same as calling `set` and then calling `save`.
    */
-  public update<K extends keyof this>(key: K, value: this[K], options?: InstanceUpdateOptions): Promise<this>;
-  public update(keys: object, options?: InstanceUpdateOptions): Promise<this>;
+  public update<K extends keyof this>(
```

Comment:
What if you want to update an attribute which also a valid property of model? This will need reviews from @SimonSchick / @papb as I am a typescript noob

Reply:
Shouldn't update just be for updating actual model data defined by the programmer?

Reply:
These are valid keys from Model class

```js
[ 'getAssociations',
  'getAssociationForAlias',
  'Mixin',
  'default',
  '_setupHooks',
  'runHooks',
  'addHook',
  'removeHook',
  'hasHook',
  'hasHooks',
  'beforeValidate',
  'afterValidate',
  'validationFailed',
  'beforeCreate',
  'afterCreate',
  'beforeDestroy',
  'afterDestroy',
  'beforeRestore',
  'afterRestore',
  'beforeUpdate',
  'afterUpdate',
  'beforeSave',
  'afterSave',
  'beforeUpsert',
  'afterUpsert',
  'beforeBulkCreate',
  'afterBulkCreate',
  'beforeBulkDestroy',
  'afterBulkDestroy',
  'beforeBulkRestore',
  'afterBulkRestore',
  'beforeBulkUpdate',
  'afterBulkUpdate',
  'beforeFind',
  'beforeFindAfterExpandIncludeAll',
  'beforeFindAfterOptions',
  'afterFind',
  'beforeCount',
  'beforeAssociate',
  'afterAssociate',
  'beforeSync',
  'afterSync',
  'beforeBulkSync',
  'afterBulkSync',
  'beforeQuery',
  'afterQuery' ]
```

This PR will prevent users from updating an model with any attribute from list above. For e.g. now you can't have a Model with attribute `default` or `hasHooks`

Reply:
Yeah I see what you are saying. Originally I thought this could actually be what we wanted since why would users be adding those attribute names to their models since they're really sequelize model functions. But, then again, they should be able to if they wanted to. So, in that sense you are right, I'll update this PR so it just limits the update keys type to `keys: Partial<this>`

Reply:
You could probably use `Exclude<keyof this, keyof Model>`.

Reply:
Associations add dynamically named methods too, you would need a way to exclude those

Reply:
Associations add dynamically named methods too, you would need a way to exclude those

Reply:
so from what @sushantdhiman mentioned, he doesn't really want to add that restriction. Basically he wants to let the user have whatever attribute names he wishes. So all I'm doing in this PR is updating the `update` signature to match what the `set` signature already had set up, which was like this:

`public set<K extends keyof this>(key: K, value: this[K], options?: SetOptions): this;`

It may not be perfect but it provides more type safety than what the current `update` signature has.

Reply:
I didn't realize that about `set`, that seems pretty bad, I mean Sequelize is probably not defined with static typing in mind, but the old DefinitelyTyped defs (don't know if they still exist, but I converted them to FlowTyped defs) actually used the `Model` type parameters to determine what attributes were available.  The current approach of the defs in this package seems destined for trouble...